feat(tests): port ethereum/tests `CALLDATALOAD` tests

LouisTsai-Csie opened the pull request:
## 🗒️ Description
<!-- Brief description of the changes introduced by this PR -->

Migrate `calldataload` test from [ethereum/tests](https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/VMTests/vmTests/calldataloadFiller.yml)

## 🔗 Related Issues
<!-- Reference any related issues using the GitHub issue number (e.g., Fixes #123) -->

[https://github.com/ethereum/execution-spec-tests/issues/972](https://t.lever-analytics.com/email-link?dest=https%3A%2F%2Fgithub.com%2Fethereum%2Fexecution-spec-tests%2Fissues%2F972&eid=b2754d1c-26f3-497b-a30c-64969bcc2825&idx=0&token=UbPKThYOl-abrWJ0Xph65E56huc)

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [ ] All: Considered squashing commits to improve commit history.
- [x] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [x] Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been added to [converted-ethereum-tests.txt](/ethereum/execution-spec-tests/blob/main/converted-ethereum-tests.txt).
- [x] Tests: A PR with removal of converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been opened.
- [x] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

winsvega commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,140 @@
+    "sstore_test_opcode,calldata_opcode,validate_length,expected_storage",
+    [
+        (
+            Bytecode(Op.SSTORE(key=0x0, value=Op.CALLDATALOAD(offset=0x0))),
```
Looks like the code remains unchanged. It can be just parametrization by offset variable

winsvega commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,140 @@
+            0x21,
+            (
+                Account(
+                    storage={
```
This seems to be expected calldataload memory. 
Can omit Account, storage defenition on each parameter

winsvega commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,140 @@
+    validation_contract = pre.deploy_contract(
+        code=(
+            Op.CALL(
+                gas=0xFFFFFF,
```
Op.GAS. and if doesn't work on Frontier
Op.Sub(Op.Gas, 100_000)

LouisTsai-Csie replied:
Could you please provide more details?

Is it because EIP-150 was not yet been implemented in the `Frontier` upgrade, causing the `CALL` instruction to forward all available gas rather than reserving some for following operations?

winsvega replied:
untill Homestead if CALL have gas=100000 it will consume all of it. 
after if CALL have gas = 100_000 it will only consume as much as it requires and will not OOG if you have lets say 50_000 and it consumed less then that.  but will fail OOG if it consumes > than you have or > 100_000

winsvega commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,140 @@
+
+@pytest.mark.parametrize(
+    "sstore_test_opcode,calldata_opcode,validate_length,expected_storage",
+    [
```
Validate_length is call_args_size

LouisTsai-Csie commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,140 @@
+    """
+    Test `CALLDATACOPY` opcode.
+
+    Based on https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/VMTests/vmTests/calldataloadFiller.yml
```
Add the corresponding [test case](https://github.com/ethereum/tests/blob/develop/GeneralStateTests/VMTests/vmTests/calldataload.json) in the ethereum-converted-tests.txt.

winsvega replied:
Have you run fill command locally?
Uv run fill tests/frontier/opcode/test...calldataload

Looks like opcode doesn't accept 20_000 notation. Always check that test fill works thats basically execute it against evm and verifies that post condition is met.

LouisTsai-Csie replied:
I did run the command on my machine previously, and the 20_000 notation works well.  My python version is `3.12.2`.

Not sure what the root cause is, and this is the screenshot for my result, there are several skipping test case, they are all for `Constantinople` tests. And the error shows `Skipped: Fork 'Constantinople' unsupported by 'ethereum-spec-evm-resolver'.`

<img width="1038" alt="Screenshot 2025-02-27 at 7 42 54 PM" src="https://github.com/user-attachments/assets/6811d025-e9a5-4789-9a36-5dc44034369b" />

LouisTsai-Csie replied:
Thank you for explanation!

LouisTsai-Csie commented:
I noticed a `TODO` in the `README` that says, "Consider squashing commits to improve commit history." When addressing issues, should I squash my commits into a single commit right away, or should I keep pushing commits until all changes are complete and then squash them into one? Which approach would be more helpful for the reviewer?

LouisTsai-Csie commented:
I analyze other PRs and found there is a specific setting for the `protected` variable. I am not sure what it is for until I analyze  related function: https://github.com/ethereum/execution-spec-tests/blob/1115574f044c2551e6fdeaaae887156f807bc07a/src/ethereum_test_types/types.py#L830

```solidity
v, r, s = (
            signature_bytes[64],
            int.from_bytes(signature_bytes[0:32], byteorder="big"),
            int.from_bytes(signature_bytes[32:64], byteorder="big"),
        )
        if self.ty == 0:
            if self.protected:
                v += 35 + (self.chain_id * 2)
            else:  # not protected
                v += 27
 ```

I want to double check, is it for [EIP-155](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md) that prevents replay attack?

winsvega commented:
Yes, protected is eip155

winsvega reviewed (changes requested):
Test looks good.
However fill doesn't work. Need to see whats wrong around hash 0x20000

marioevz commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,124 @@
+    calldataload_offset: int,
+    memory_preset_code: Bytecode,
+    call_args_size: int,
+    expected_calldataload_memory: Account,
```
```suggestion
    expected_calldataload_memory: int,
```
Correct type seems to be `int`, based on the values in your `pytest.mark.parametrize`.

marioevz commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,124 @@
+    calldata_contract = pre.deploy_contract(
+        code=(
+            memory_preset_code
+            + Op.CALL(
+                gas=Op.SUB(Op.GAS, 20_000),
+                address=sstore_contract,
+                value=0x0,
+                args_offset=0x0,
+                args_size=call_args_size,
+                ret_offset=0x0,
+                ret_size=0x0,
+            )
```
```suggestion
            + Op.CALL(
                gas=Op.SUB(Op.GAS, 20_000),
                address=sstore_contract,
                args_size=call_args_size,
            )
```
All arguments with zero value can be skipped (Except for `gas`, but not the case here).

Same comment applies to `Op.CALL` below.

marioevz commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,124 @@
+    )
+
+    tx = Transaction(
+        data=b"\x01",
```
This data is not actually used right now, so we can remove it.

_But..._ we could write a second test where test `Op.CALLDATALOAD` on the first call stack level:

Right now we are testing on the second call stack level, simply because we are using the opcode on a second contract via `Op.CALL`.

Another test could use this transaction field to test `Op.CALLDATALOAD` using a single contract.

marioevz commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,124 @@
+    tx = Transaction(
+        data=b"\x01",
+        gas_limit=100_000,
+        gas_price=0x0A,
```
```suggestion
```

Can be removed.

LouisTsai-Csie replied:
I removed it, but I wonder when will we are required to configure the value

and I saw different version across test cases:

https://github.com/ethereum/execution-spec-tests/blob/8513a8e14b62394cbebb724b6fa9a6a124ac0b7d/tests/frontier/opcodes/test_dup.py#L69

https://github.com/ethereum/execution-spec-tests/blob/8513a8e14b62394cbebb724b6fa9a6a124ac0b7d/tests/frontier/opcodes/test_calldatacopy.py#L170

https://github.com/ethereum/execution-spec-tests/blob/8513a8e14b62394cbebb724b6fa9a6a124ac0b7d/tests/frontier/opcodes/test_call_and_callcode_gas_calculation.py#L126

And also, for the `gas_limit` value:

https://github.com/ethereum/execution-spec-tests/blob/8513a8e14b62394cbebb724b6fa9a6a124ac0b7d/tests/frontier/opcodes/test_call_and_callcode_gas_calculation.py#L125

https://github.com/ethereum/execution-spec-tests/blob/8513a8e14b62394cbebb724b6fa9a6a124ac0b7d/tests/frontier/opcodes/test_calldatacopy.py#L169

https://github.com/ethereum/execution-spec-tests/blob/8513a8e14b62394cbebb724b6fa9a6a124ac0b7d/tests/frontier/opcodes/test_dup.py#L68

The `gas_limit` and `gas_price` value seems irrelevant to the test case

marioevz commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,124 @@
+        gas_price=0x0A,
+        protected=False if fork in [Frontier, Homestead] else True,
+        to=validation_contract,
+        value=0x01,
```
```suggestion
```

Also can be removed.

marioevz reviewed (commented):
Hi Louis! Thanks for the PR, it looks pretty good!

Just a couple of comments though.

danceratopz commented:
> Test looks good. However fill doesn't work. Need to see whats wrong around hash 0x20000

Hey @LouisTsai-Csie, thanks for the PR! Just a heads up, the failing Github Action [Evmone Coverage Report](https://github.com/ethereum/execution-spec-tests/actions/runs/13526145990/job/37802316036?pr=1248) is a separate issue unrelated to your test. It should be fixed with #1268.

LouisTsai-Csie commented on tests/frontier/opcodes/test_calldataload.py:
```diff
@@ -0,0 +1,124 @@
+    Test `CALLDATACOPY` opcode.
+
+    Based on https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/VMTests/vmTests/calldataloadFiller.yml
+    """
```
I update the test id and the description according to the test cases, it should be more clear now

two_bytes -> calldata_short_start_0
word_n_byte -> calldata_sufficient_with_offset
34_bytes -> calldata_partial_word_at_offset

marioevz replied:
I should have been a bit more clear with my comment sorry!

What I meant was to document the test function parameters, so 'calldataload_offset', 'memory_preset_code' and so on.

I think we can leave the test descriptions you wrote since they are indeed helpful (we should actually do this more often) and on top of that also describe the test function parameters.

LouisTsai-Csie commented:
@marioevz Thanks for your review, I updated the file and left several comments. 

I wonder if I should use `git reset` to clean up my commit history and ensure that each commit represents an independent task. Based on my previous experience, some teams prefer not to reset commit history. I’d love to hear your thoughts on this.

marioevz commented:
> @marioevz Thanks for your review, I updated the file and left several comments. 
> 
> 
> 
> I wonder if I should use `git reset` to clean up my commit history and ensure that each commit represents an independent task. Based on my previous experience, some teams prefer not to reset commit history. I’d love to hear your thoughts on this.

We don't have strict guidelines regarding how to merge commits in this repository, but usually we simply squash and merge a PR, which means the commit history is not normally preserved.

For PRs containing only test changes like this one, squashing commits would be the approach, so don't worry too much about preserving every commit 👍 

An example of a PR we would like to preserve the history is one where not only tests are affected but  it also contains framework changes.

marioevz reviewed (commented):
One last minor comment and we should be able to merge. Thanks!

LouisTsai-Csie commented:
I've updated the comment. Please let me know if it is clear. Thanks!

winsvega commented:
need to check why this line coverage is lost
```
 evmone::state::MPTNode::ext(evmone::state::(anonymous namespace)::Path const&, std::unique_ptr<evmone::state::MPTNode, std::default_delete<evmone::state::MPTNode> >)
evmone::state::State::journal_balance_change(evmc::address const&, intx::uint<256u> const&)
```

the other seem to be related to add opcode not being called anymore

danceratopz commented:
Hey @LouisTsai-Csie! I'd recommend rebasing this to `main` and then we should look to get this merged!

> @marioevz Thanks for your review, I updated the file and left several comments.
> 
> I wonder if I should use `git reset` to clean up my commit history and ensure that each commit represents an independent task. Based on my previous experience, some teams prefer not to reset commit history. I’d love to hear your thoughts on this.

Just FYI: We just added some guidelines in CONTRIBUTING.md:
https://github.com/ethereum/execution-spec-tests/blob/main/CONTRIBUTING.md#pr-expectations

It's available in our online pages; you can also find some docs on coding standards here:
https://eest.ethereum.org/main/getting_started/contributing/

We're not too precious about the git history within a PR. Although the more logical the commits, the better, of course. As-is, I think this PR is very understandable, so no need to re-commit your changes here :+1:

danceratopz commented:
@LouisTsai-Csie ah, I was too fast. Someone else also ported `CALLDATALOAD` in the meantime :cry: We should compare implementations before you rebase.

The other version was implemented here: 
- https://github.com/ethereum/execution-spec-tests/issues/1236
